File: verify-links.mjs
#!/usr/bin/env node
/**
 * verify-links.mjs
 *
 * For each wikiLink in the static/content/ JSON files, extracts the
 * surrounding Suetonius context, queries Wikipedia with "Suetonius"
 * appended to disambiguate, and compares the linked article vs the
 * suggested article. Outputs a markdown table and emails it.
 */

import { readFileSync, readdirSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { createRequire } from 'module';
import smtplib from 'smtplib';
import { MIMEText, MIMEMultipart } from 'mime';

const __dirname = dirname(fileURLToPath(import.meta.url));
const require = createRequire(import.meta.url);
const WIKIPEDIA_API = 'https://en.wikipedia.org/w/api.php';
const CONTENT_DIR = join(__dirname, 'static', 'content');
const OUTPUT_PATH = join(__dirname, 'verify_links.md');

// ── Wikipedia helpers ───────────────────────────────────────────────────────────

async function wikipediaSearch(query) {
  const url = `${WIKIPEDIA_API}?action=query&list=search&srsearch=${encodeURIComponent(query)}&format=json&srlimit=1`;
  const resp = await fetch(url);
  const data = await resp.json();
  const hits = data?.query?.search ?? [];
  if (!hits.length) return null;
  return hits[0].title;
}

function extractTitleFromUrl(url) {
  if (!url) return '';
  return decodeURIComponent(url.split('/').pop().replace(/_/g, ' '));
}

// ── Context extraction ────────────────────────────────────────────────────────

function getContext(text, location, radius = 80) {
  const start = Math.max(0, location - radius);
  const end = Math.min(text.length, location + radius);
  let excerpt = text.slice(start, end).replace(/\n/g, ' ').replace(/\s+/g, ' ');
  if (start > 0) excerpt = '...' + excerpt;
  if (end < text.length) excerpt = excerpt + '...';
  return excerpt;
}

// ── Load all emperor data ─────────────────────────────────────────────────────

function loadAllEmperors() {
  const files = readdirSync(CONTENT_DIR).filter(f => f.endsWith('.json'));
  const emperors = [];
  for (const file of files) {
    const slug = file.replace('.json', '');
    const data = JSON.parse(readFileSync(join(CONTENT_DIR, file), 'utf-8'));
    emperors.push({ slug, data });
  }
  return emperors;
}

function collectLinksWithContext(emperors) {
  const collected = new Map(); // name -> { url, contexts[], slug }

  for (const { slug, data } of emperors) {
    const allText = data.sections.map(s => s.en).join('\n');

    for (const section of data.sections) {
      const text = section.en;
      const wikiLinks = section.wikiLinks ?? {};

      for (const [name, url] of Object.entries(wikiLinks)) {
        if (!name || !url) continue;

        // Find all occurrences of name in the full emperor text
        const regex = new RegExp(`\\b${name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'gi');
        let match;
        const contexts = [];
        while ((match = regex.exec(text)) !== null) {
          contexts.push(getContext(text, match.index));
        }

        if (!collected.has(name)) {
          collected.set(name, { url, contexts: [], slug });
        } else {
          // Merge contexts, keep unique
          const existing = collected.get(name);
          for (const ctx of contexts) {
            if (!existing.contexts.includes(ctx)) existing.contexts.push(ctx);
          }
        }
      }
    }
  }

  return collected;
}

// ── Markdown builder ──────────────────────────────────────────────────────────

function buildMarkdown(results) {
  const lines = [
    '# Suetonius Links — Disambiguation Verification\n',
    '*Powered by Wikipedia search with "Suetonius" appended for disambiguation*\n',
    '## How to read this\n',
    '- **Linked Article** = what we currently have in the JSON\n',
    '- **Suggested Article** = top Wikipedia result when searching "Name Suetonius" + context\n',
    '- **Match?** = ✓ if they agree, ✗ if they diverge\n',
    '---\n',
    '| Name | Context (first occurrence) | Linked Article | Suggested (Suetonius) | Match? |',
    '|-|<-|<-|<-|-:|',
  ];

  for (const r of results) {
    const ctx = r.context.length > 120 ? r.context.slice(0, 120) + '...' : r.context;
    lines.push(
      `| ${r.name} | ${ctx} | ${r.linkedTitle} | ${r.suggestedTitle} | ${r.match} |`
    );
  }

  const stats = {
    total: results.length,
    match: results.filter(r => r.match === '✓').length,
    mismatch: results.filter(r => r.match === '✗').length,
    uncertain: results.filter(r => r.match === '?').length,
  };

  lines.push(
    '\n## Summary\n',
    `- **Total unique links verified:** ${stats.total}`,
    `- **✓ Matches:** ${stats.match}`,
    `- **✗ Mismatches:** ${stats.mismatch}`,
    `- **? Uncertain:** ${stats.uncertain}`,
  );

  if (stats.mismatch > 0) {
    lines.push(
      '\n## Mismatches to review\n',
      ...results
        .filter(r => r.match === '✗')
        .map(r => `- **${r.name}** → currently "${r.linkedTitle}" but Wikipedia+Suetonius suggests "${r.suggestedTitle}"`)
    );
  }

  return lines.join('\n');
}

// ── Email ─────────────────────────────────────────────────────────────────────

async function sendEmail(markdownContent, toEmail) {
  const msg = new MIMEMultipart();
  msg['Subject'] = 'Suetonius Links — Disambiguation Verification';
  msg['From'] = 'bushap5000@gmail.com';
  msg['To'] = toEmail;

  const body = (
    'Verification report attached. Check the Match column:\n'
    + '- ✓ = linked article matches Wikipedia+Suetonius disambiguation\n'
    + '- ✗ = mismatch, review needed\n'
    + '- ? = no suggestion found\n\n'
    + 'Run by your frontier model for a second opinion.\n'
    + 'Pay special attention to ✗ rows — those are likely wrong article picks.'
  );
  msg.attach(new MIMEText(body, 'plain'));

  const attachment = new MIMEText(markdownContent, 'markdown');
  attachment.addHeader('Content-Disposition', 'attachment', filename='verify_links.md');
  msg.attach(attachment);

  const GMAIL_USER = 'bushap5000@gmail.com';
  const GMAIL_APP_PASSWORD = 'chchmqtghfiptkte';

  const server = smtplib.SMTP('smtp.gmail.com', 587);
  server.setDebugLevel(0);
  server.connect();
  server.starttls();
  server.login(GMAIL_USER, GMAIL_APP_PASSWORD);
  server.sendMessage(msg);
  server.quit();

  console.log(`Email sent to ${toEmail}`);
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  console.log('Loading emperor data...');
  const emperors = loadAllEmperors();
  console.log(`Loaded ${emperors.length} emperors`);

  console.log('Collecting wikiLinks...');
  const collected = collectLinksWithContext(emperors);
  console.log(`Found ${collected.size} unique linked names`);

  const results = [];
  let i = 0;

  for (const [name, { url, contexts, slug }] of collected.entries()) {
    i++;
    const linkedTitle = extractTitleFromUrl(url);
    const context = contexts[0] ?? '';

    // Build search query: name + "Suetonius" + short context snippet
    const contextSnippet = context.replace(/[^a-zA-Z0-9 ]/g, ' ').replace(/\s+/g, ' ').trim().slice(0, 60);
    const searchQuery = contextSnippet
      ? `${name} Suetonius ${contextSnippet}`
      : `${name} Suetonius`;

    console.log(`[${i}/${collected.size}] ${name} → searching "${searchQuery}"`);

    const suggestedTitle = await wikipediaSearch(searchQuery);

    let match = '?';
    if (suggestedTitle && linkedTitle) {
      const normLinked = linkedTitle.toLowerCase().replace(/\s*\(\w+\)/g, '').replace(/[_-]/g, ' ');
      const normSuggested = suggestedTitle.toLowerCase().replace(/\s*\(\w+\)/g, '').replace(/[_-]/g, ' ');
      match = normLinked === normSuggested ? '✓' : '✗';
    } else if (!suggestedTitle) {
      match = '? (no wiki suggestion)';
    }

    results.push({ name, context, linkedTitle, suggestedTitle: suggestedTitle || '(none)', match });
  }

  console.log('\nBuilding markdown...');
  const md = buildMarkdown(results);

  writeFileSync(OUTPUT_PATH, md);
  console.log(`Markdown saved to ${OUTPUT_PATH}`);

  const mismatches = results.filter(r => r.match === '✗').length;
  console.log(`\nResults: ${results.length} total, ${mismatches} mismatches`);

  if (mismatches > 0) {
    console.log('\nMismatches:');
    for (const r of results.filter(r => r.match === '✗')) {
      console.log(`  ✗ ${r.name}: linked "${r.linkedTitle}" vs suggested "${r.suggestedTitle}"`);
    }
  }

  console.log('\nSending email...');
  await sendEmail(md, 'shapiro.peter@gmail.com');
  console.log('Done.');
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});

File: verify-links.py
#!/usr/bin/env python3
"""
verify-links.py

For each wikiLink in static/content/*.json, extracts surrounding context,
queries Wikipedia to verify the linked article is the correct historical figure,
and compares the linked article vs the suggested article. Emits a markdown
table and emails it as an attachment to shapiro.peter@gmail.com.

Peter's correction (2026-03-26): The "Name Suetonius" search heuristic generates
massive false negatives — Wikipedia search returns articles ABOUT Suetonius the
author rather than the historical figure being referenced. The correct approach
is to use context (emperor name, chapter heading) to disambiguate, not append
"Suetonius" to every query.
"""

import json
import re
import smtplib
import os
from pathlib import Path
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.mime.base import MIMEBase
from email import encoders

import requests

# ── Config ────────────────────────────────────────────────────────────────────

CONTENT_DIR = Path(__file__).parent / "static" / "content"
OUTPUT_PATH = Path(__file__).parent / "verify_links.md"
WIKIPEDIA_API = "https://en.wikipedia.org/w/api.php"
GMAIL_USER = "bushap5000@gmail.com"
GMAIL_APP_PASSWORD = "***"
RECIPIENT = "shapiro.peter@gmail.com"

# Emperor-specific context terms for disambiguation
# Maps emperor slug → terms to include in Wikipedia search to narrow results
EMPEROR_CONTEXT = {
    "julius": "Roman Republic dictator",
    "augustus": "Roman emperor first",
    "tiberius": "Roman emperor second",
    "caligula": "Roman emperor third",
    "claudius": "Roman emperor fourth",
    "nero": "Roman emperor fifth",
    "galba": "Roman emperor sixth",
 "otho": "Roman emperor seventh",
    "vitellius": "Roman emperor eighth",
    "vespasian": "Roman emperor ninth",
    "titus": "Roman emperor tenth",
    "domitian": "Roman emperor eleventh",
}

# Known ambiguous names that need disambiguation even with emperor context
# Format: display_name -> (wrong_article_prefix, correct_context_hint)
AMBIGUOUS_NAMES = {
    "Lepidus": ("Marcus Aemilius Lepidus", "triumvir not later consul"),
    "Brutus": ("Marcus Junius Brutus", "Republic assassin not other Brutus"),
    "Pompey": ("Gnaeus Pompeius Magnus", "Roman general not other Pompey"),
    "Cato": ("Marcus Porcius Cato", "Republic statesman not other Cato"),
    "Sulla": ("Lucius Cornelius Sulla", "Roman dictator not other Sulla"),
    "Cicero": ("Marcus Tullius Cicero", "Roman orator not other Cicero"),
    "Marius": ("Gaius Marius", "Roman general not other Marius"),
    "Crassus": ("Marcus Licinius Crassus", "triumvir not other Crassus"),
}


# ── Wikipedia helpers ──────────────────────────────────────────────────────────

def wikipedia_opensearch(query: str) -> list[str]:
    """Return Wikipedia title suggestions using opensearch (prefix search).
    
    Uses the Wikipedia OpenSearch API which returns title matches rather
    than full-text search matches. Best for disambiguating Roman names
    when you know the rough correct form.
    """
    params = {
        "action": "opensearch",
        "search": query,
        "limit": 5,
        "namespace": 0,
        "format": "json",
    }
    headers = {
        "User-Agent": "TreadwellBot/1.0 (Suetonius-link-verification; mailto:bushap@gmail.com)"
    }
    try:
        resp = requests.get(WIKIPEDIA_API, params=params, headers=headers, timeout=10)
        data = resp.json()
        return data[1] if isinstance(data, list) and len(data) > 1 else []
    except Exception:
        return []


def wikipedia_srsearch(query: str) -> str | None:
    """Return top Wikipedia article title using full-text search.
    
    Uses the Wikipedia search API which returns articles matching the query
    text. Good for names that are specific enough to find the right article
    on their own (Cleopatra, Pompey, geographic places).
    Peter correction: Do NOT append "Suetonius" — it causes search to return
    articles about the author rather than the historical figure.
    """
    params = {
        "action": "query",
        "list": "search",
        "srsearch": query,
        "format": "json",
        "srlimit": 1,
    }
    headers = {
        "User-Agent": "TreadwellBot/1.0 (Suetonius-link-verification; mailto:bushap@gmail.com)"
    }
    try:
        resp = requests.get(WIKIPEDIA_API, params=params, headers=headers, timeout=10)
        data = resp.json()
        hits = data.get("query", {}).get("search", [])
        return hits[0]["title"] if hits else None
    except Exception:
        return None


# Names that are short and ambiguous — need hints from AMBIGUOUS_NAMES
_SHORT_AMBIGUOUS = frozenset(k for k in AMBIGUOUS_NAMES)


def wikipedia_suggest(name: str, emperor_slug: str) -> str | None:
    """Get the best Wikipedia article suggestion for a name.
    
    Strategy (Peter correction — do NOT append "Suetonius"):
    1. Short ambiguous names (Brutus, Lepidus, Cato, etc.) → use opensearch
       with the full name + disambiguation hint from AMBIGUOUS_NAMES
    2. Full/disambiguated names (Ptolemy XIII, Gaius Marius) → opensearch directly
    3. Common names and places (Cleopatra, Pompey, Gaul) → srsearch with emperor context
    
    This avoids the false-negative problem where "Name Suetonius" search
    returns articles about the author Suetonius.
    """
    # Case 1: short ambiguous names — use hints
    if name in _SHORT_AMBIGUOUS:
        correct_name, hint = AMBIGUOUS_NAMES[name]
        suggestions = wikipedia_opensearch(f"{correct_name} {hint}")
        return suggestions[0] if suggestions else None
    
    # Case 2: names that look already disambiguated — try opensearch directly
    if any(c in name for c in "(,)"):
        suggestions = wikipedia_opensearch(name)
        return suggestions[0] if suggestions else None
    
    # Case 3: common names and places — use srsearch with emperor context
    emperor_context = EMPEROR_CONTEXT.get(emperor_slug, "")
    query = f"{name} {emperor_context}" if emperor_context else name
    return wikipedia_srsearch(query)


def title_from_url(url: str) -> str:
    """Extract article title from a Wikipedia URL."""
    if not url:
        return ""
    return url.split("/")[-1].replace("_", " ")


# ── Context extraction ─────────────────────────────────────────────────────────

def get_context(text: str, location: int, radius: int = 80) -> str:
    start = max(0, location - radius)
    end = min(len(text), location + radius)
    excerpt = text[start:end].replace("\n", " ").replace(r"\s+", " ")
    if start > 0:
        excerpt = "..." + excerpt
    if end < len(text):
        excerpt = excerpt + "..."
    return excerpt.strip()


# ── Load all emperor data ──────────────────────────────────────────────────────

def load_all_emperors():
    files = list(CONTENT_DIR.glob("*.json"))
    emperors = []
    for f in files:
        slug = f.stem
        data = json.loads(f.read_text())
        emperors.append({"slug": slug, "data": data})
    return emperors


def collect_links_with_context(emperors):
    """Collect all unique wikiLinks across all emperors, with context."""
    collected = {}  # name -> {url, contexts, slug}

    for emperor in emperors:
        data = emperor["data"]
        sections = data.get("sections", [])
        slug = emperor["slug"]

        for section in sections:
            text = section.get("en", "")
            wiki_links = section.get("wikiLinks", {})

            for name, url in wiki_links.items():
                if not name or not url:
                    continue

                try:
                    pattern = re.compile(r"\b" + re.escape(name) + r"\b", re.IGNORECASE)
                except re.error:
                    continue

                contexts = []
                for m in pattern.finditer(text):
                    contexts.append(get_context(text, m.start()))

                if name not in collected:
                    collected[name] = {"url": url, "contexts": contexts, "slug": slug}
                else:
                    existing = collected[name]
                    for ctx in contexts:
                        if ctx not in existing["contexts"]:
                            existing["contexts"].append(ctx)

    return collected


# ── Markdown builder ──────────────────────────────────────────────────────────

def build_markdown(results):
    lines = [
        "# Suetonius Links — Disambiguation Verification\n",
        "*Using Wikipedia title search with emperor historical context (NOT 'Suetonius' suffix)*\n",
        "## How to read this\n",
        "- **Linked Article** = what we currently have in the JSON\n",
        "- **Suggested Article** = top Wikipedia title match using emperor context disambiguation\n",
        "- **Match?** = ✓ if they agree, ✗ if they diverge, ? if no suggestion found\n",
        "---\n",
        "| N | Name | Context | Linked Article | Suggested | Match? |\n",
        "|--:|--|--|--|--|--:|\n",
    ]

    for i, r in enumerate(results, 1):
        ctx = r["context"]
        if len(ctx) > 100:
            ctx = ctx[:100] + "..."
        lines.append(
            f"| {i} | {r['name']} | {ctx} | {r['linked_title']} | "
            f"{r['suggested_title']} | {r['match']} |"
        )

    stats = {
        "total": len(results),
        "match": sum(1 for r in results if r["match"] == "✓"),
        "mismatch": sum(1 for r in results if r["match"] == "✗"),
        "uncertain": sum(1 for r in results if r["match"] == "?"),
    }

    lines.extend([
        "\n## Summary\n",
        f"- **Total unique links verified:** {stats['total']}\n",
        f"- **✓ Matches:** {stats['match']}\n",
        f"- **✗ Mismatches:** {stats['mismatch']}\n",
        f"- **? Uncertain:** {stats['uncertain']}\n",
    ])

    mismatches = [r for r in results if r["match"] == "✗"]
    if mismatches:
        lines.append("\n## Mismatches to review\n")
        for r in mismatches:
            lines.append(
                f"- **{r['name']}** → currently linked to \"{r['linked_title']}\" "
                f"but Wikipedia suggests \"{r['suggested_title']}\"\n"
            )

    return "\n".join(lines)


# ── Email ─────────────────────────────────────────────────────────────────────

def send_email(markdown_content: str, to_email: str):
    msg = MIMEMultipart()
    msg["Subject"] = "Suetonius Links — Disambiguation Verification"
    msg["From"] = GMAIL_USER
    msg["To"] = to_email

    body = (
        "Verification report attached. Check the Match column:\n"
        "- \\u2713 = linked article matches Wikipedia disambiguation\n"
        "- \\u2717 = mismatch, review needed\n"
        "- ? = no suggestion found\n\n"
        "Uses emperor historical context for disambiguation (NOT 'Suetonius' suffix).\n"
        "Run by your frontier model for a second opinion.\n"
        f"Total: {len([l for l in markdown_content.splitlines() if l.startswith('|') and not l.startswith('|--')])} rows"
    )
    msg.attach(MIMEText(body, "plain"))

    part = MIMEBase("application", "octet-stream")
    part.set_payload(markdown_content.encode("utf-8"))
    encoders.encode_base64(part)
    part.add_header("Content-Disposition", "attachment", filename="verify_links.md")
    msg.attach(part)

    with smtplib.SMTP("smtp.gmail.com", 587) as server:
        server.starttls()
        server.login(GMAIL_USER, GMAIL_APP_PASSWORD)
        server.send_message(msg)

    print(f"Email sent to {to_email}")


# ── Main ─────────────────────────────────────────────────────────────────────

def main():
    print("Loading emperor data...")
    emperors = load_all_emperors()
    print(f"Loaded {len(emperors)} emperors")

    print("Collecting wikiLinks...")
    collected = collect_links_with_context(emperors)
    print(f"Found {len(collected)} unique linked names")

    results = []
    for i, (name, info) in enumerate(collected.items(), 1):
        url = info["url"]
        contexts = info["contexts"]
        slug = info["slug"]

        linked_title = title_from_url(url)
        context = contexts[0] if contexts else ""

        # Use proper Wikipedia disambiguation (Peter's correction — no "Suetonius" suffix)
        suggested_title = wikipedia_suggest(name, slug)
        print(f"[{i}/{len(collected)}] {name} (emperor={slug}) -> suggested: {suggested_title}")

        # Check match: normalize titles for comparison
        if suggested_title and linked_title:
            linked_slug = linked_title.lower().replace(" ", "_").replace("-", "_")
            linked_slug = re.sub(r"[()]", "", linked_slug)
            suggested_slug = suggested_title.lower().replace(" ", "_").replace("-", "_")
            suggested_slug = re.sub(r"[()]", "", suggested_slug)

            slug_contained = (linked_slug in suggested_slug) or (suggested_slug in linked_slug)
            linked_words = set(re.sub(r"[^a-z]", "", w.lower()) for w in linked_title.split() if len(w) > 2)
            suggested_words = set(re.sub(r"[^a-z]", "", w.lower()) for w in suggested_title.split() if len(w) > 2)
            word_overlap = len(linked_words & suggested_words)
            min_overlap = max(2, len(linked_words) // 2)
            match = "✓" if slug_contained or word_overlap >= min_overlap else "✗"
        elif suggested_title:
            match = "?"
        else:
            match = "? (no wiki)"

        results.append({
            "name": name,
            "context": context,
            "linked_title": linked_title,
            "suggested_title": suggested_title or "(none)",
            "match": match,
        })

    print("\nBuilding markdown...")
    md = build_markdown(results)
    OUTPUT_PATH.write_text(md)
    print(f"Markdown saved to {OUTPUT_PATH}")

    mismatches = [r for r in results if r["match"] == "✗"]
    print(f"\nResults: {len(results)} total, {len(mismatches)} mismatches")
    for r in mismatches:
        print(f"  ✗ {r['name']}: linked '{r['linked_title']}' vs suggested '{r['suggested_title']}'")

    print("\nSending email...")
    send_email(md, RECIPIENT)
    print("Done.")


if __name__ == "__main__":
    main()
